src/main/java/org/elasticsearch/common/util/concurrent/BaseFuture.java

```diff
@@ -118,6 +117,22 @@ public V get() throws InterruptedException, ExecutionException {
         return sync.get();
     }
 
+    /**
+     * Similar to {@link #get()} with the exception that it never blocks. If the operation wasn't completed yet
+     * it will return {@code null}. Otherwise it will behave in a similar fashion to {@link #get()} and will return
+     * the computed result, including throwing a {@link CancellationException} or {@link ExecutionException} if
+     * the computation was cancelled or threw an error (respectively).
+     *
+     * @throws CancellationException if the computation was cancelled
+     * @throws ExecutionException    if the computation threw an exception
+     */
+    public V tryGet() throws CancellationException, ExecutionException {
+        if (sync.isDone() == false) {
+            return null;
+        }
+        return sync.getValue();
+    }
```

Comment:
yeah, I considered a couple of approaches:

- The one above (overloading the `null` value to indicate not done). 
- Having a special case for doing get with timeout set to 0 where we don't assert on a networking thread. 
- Having a method like you suggest called something like `getExpectingDone` which is expected to be call only after the future is guaranteed to be done. If feel `tryGet` sounds like something very light where the expectation is that is not done (i.e., a syntactic sugar for if (isDone) get() ).

I don't like any of the above much, but I had a slight preference to the first. If you like the second better, I can do that. Of course, new ideas are welcome :)